src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowJob.java

```diff
@@ -339,12 +343,27 @@ public String getShortDescription() {
 
     @Exported
     @Override public boolean isConcurrentBuild() {
-        return !Boolean.FALSE.equals(concurrentBuild);
+        DisableConcurrentBuildsJobProperty p = getProperty(DisableConcurrentBuildsJobProperty.class);
+        if (p != null) {
+            return false;
+        } else {
+            /* settings compatibility */
+            return !Boolean.FALSE.equals(concurrentBuild);
+        }
     }
-    
+
     public void setConcurrentBuild(boolean b) throws IOException {
-        concurrentBuild = b ? null : false;
-        save();
+        concurrentBuild = null;
+        BulkChange bc = new BulkChange(this);
+        try {
+            removeProperty(DisableConcurrentBuildsJobProperty.class);
+            if (!b) {
+                addProperty(new DisableConcurrentBuildsJobProperty());
```

Comment:
Adding such an optimization with a few line comment explaining what exactly the logic is for the optimization, since it's a bit weird at first glance. =)